### PR TITLE
fix: Add epoch fallback for MIP-04 media decryption

### PR DESCRIPTION
**Pull request by JSKitty**
## Summary

*(Edited based on the latest commit cd37f5179d05af19493cf7a453de39a7eb9bf125)*

**Problem:**
In Vector, MLS attachments shared at older epochs failed to decrypt because MDK only tried the current epoch's exporter secret for key derivation. Once the group epoch advanced (from member joins/leaves/updates), the original key was no longer reachable; if a member needed to re-download an older attachment, it would fail to decrypt with no possible workaround.

**Solution:**
`decrypt_from_download` now resolves the correct decryption key via an O(1) epoch hint lookup. Every processed message already stores its epoch alongside the IMETA tag (which contains the file's `x <sha256_hex>` hash), so we query that directly to find which epoch's exporter secret to use for key derivation.

## Decryption Flow

```
decrypt_from_download(encrypted_data, reference)
│
│  SQL: SELECT epoch FROM messages
│    WHERE group_id = ? AND tags LIKE '%x <sha256_hex>%'
│    AND epoch IS NOT NULL LIMIT 1
│         │
│  get_group_exporter_secret(epoch)
│         │
│  derive_encryption_key_with_secret(secret, ...)
│         │
│  ChaCha20-Poly1305 decrypt with AAD
│         │
│  SHA256(plaintext) == original_hash? → Ok / Err
```

One SQL query, one secret fetch, one key derivation, one decryption, one hash check.

## Benchmark

Tested on a 45kb image ~2 epochs behind: **8.9ms** (44% faster than the previous approach in the original commit). The improvement scales with epoch count and file size.

## Key Changes

- **`mdk-storage-traits/messages/mod.rs`**: Added `find_message_epoch_by_tag_content` to `MessageStorage` trait — searches for a message whose serialized tags contain a given substring and returns its epoch.

- **`mdk-sqlite-storage/messages.rs`**: SQLite implementation using `SELECT epoch FROM messages WHERE mls_group_id = ? AND tags LIKE ? AND epoch IS NOT NULL LIMIT 1`.

- **`mdk-memory-storage/messages.rs`**: Memory implementation iterating group messages, serializing tags to JSON, checking for substring match.

- **`mdk-core/encrypted_media/manager.rs`**: Added `try_decrypt_with_epoch_hint` which looks up epoch via `find_message_epoch_by_tag_content` using `x <hex_hash>` as search term, fetches that epoch's exporter secret, derives key, and decrypts. `decrypt_from_download` delegates directly to this method.

## Security Impact

None. Encryption scheme, key derivation, AAD construction, and hash verification are unchanged. The search term is passed as a parameterized SQL bind.

## Tests

- All 22 encrypted_media::manager tests pass
- `cargo test -p mdk-sqlite-storage` — 172 tests pass
- `cargo test -p mdk-memory-storage` — 95 tests pass
- `cargo clippy -p mdk-core --features mip04` — 0 warnings
- Manual implementation testing in the Vector Client - decrypting old attachments - receiving new attachments.

⚠️ Security-sensitive changes

This PR adds an epoch-hint lookup to MIP-04 media decryption so attachments encrypted under a previous MLS epoch can be decrypted using the exporter secret from the epoch when the attachment was published, avoiding brute-force epoch attempts and fixing failures decrypting older attachments. decrypt_from_download now locates a processed message whose tags contain the IMETA file hash, uses that message's epoch to fetch the historical exporter secret, derives the encryption key for that epoch, decrypts with ChaCha20-Poly1305 using the same AAD, and verifies SHA256(plaintext) matches the original hash.

What changed:
- decrypt_from_download now tries an O(1) epoch-hint path (try_decrypt_with_epoch_hint) and falls back to the current-epoch exporter-secret path on failure.
- Added derive_encryption_key_with_secret(pub(crate)) to derive an encryption key directly from a provided exporter secret for historical-epoch decryption.
- Added MessageStorage::find_message_epoch_by_tag_content and implemented it for SQLite (parameterized SELECT epoch FROM messages WHERE tags LIKE ? with proper escaping) and in-memory storage (serialize tags to JSON and substring-match).
- Added decrypt_and_verify helper to perform ChaCha20-Poly1305 decryption with AAD and SHA-256 plaintext verification.
- Introduced EncryptedMediaError::NoExporterSecretForEpoch(u64) to indicate a resolved epoch lacks an exporter secret.
- Affected crates: mdk-core, mdk-storage-traits, mdk-sqlite-storage, mdk-memory-storage.

Security impact:
- No changes to the encryption algorithm, nonce construction, AAD formation, or HKDF-SHA256 derivation semantics; key derivation now accepts an explicit exporter secret but preserves prior HKDF context strings.
- Key material continues to be handled as Secret<[u8; 32]> and is not persisted differently.
- SQLite lookup uses parameterized queries with escaping to mitigate injection risks; in-memory storage uses JSON substring matching.
- The new error variant exposes an epoch identifier (NoExporterSecretForEpoch(u64)), so callers should treat epoch identifiers as potentially sensitive when logging or propagating errors.

Protocol changes:
- Implements MIP-04 epoch-fallback behavior for media decryption by resolving historical exporter secrets via stored IMETA tags; there are no changes to MLS key schedule, message formats, or MIP semantics beyond the fallback lookup.

API surface:
- New trait method: MessageStorage::find_message_epoch_by_tag_content(group_id, content_substring) -> Result<Option<u64>, MessageError>.
- New crate-visible function: pub(crate) fn derive_encryption_key_with_secret(...) in mdk-core.
- New public enum variant: EncryptedMediaError::NoExporterSecretForEpoch(u64).
- No storage schema changes (reuses existing messages.tags column) and no UniFFI/FFI boundary changes.

Testing:
- Tests added/extended for epoch-hint success, epoch fallback, error propagation, and storage edge cases; encrypted_media::manager, mdk-sqlite-storage, and mdk-memory-storage test suites pass locally.
- Benchmarks: a ~45 KB image ~2 epochs behind decrypts in ~8.9 ms (~44% faster than the prior approach), with improvement scaling by epoch count and file size.
- Cargo clippy for mdk-core with mip04 feature reported 0 warnings and coverage was validated during review.

### CODE REVIEW

**Comment by JSKitty** [comment]
I realised there's a more efficient way to do this (the 'epoch brute-forcing' is wasteful in long-standing groups or with large files), going to change this to use lookups with our existing MDK storage, give me a little bit.

**Comment by dannym-arx** [comment]
thank you so much for this contribution, we'll review the code thoroughly and probably get this merged! but from a quick skim it looks fine to me.

Just a quick thing, please [sign your commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and force push with lease, we cannot merge this until you do

File: crates/mdk-core/src/encrypted_media/manager.rs [hunk 1]
```diff
@@ -163,6 +172,57 @@ where
...
+            .ok_or(EncryptedMediaError::DecryptionFailed {
+                reason: "No epoch hint found for media reference".to_string(),
+            })?;
+
+        let secret = self
+            .mdk
+            .storage()
+            .get_group_exporter_secret(&self.group_id, epoch)
+            .map_err(|_| EncryptedMediaError::DecryptionFailed {
+                reason: format!("Failed to read exporter secret for hinted epoch {}", epoch),
```
Comment (dannym-arx) [inline crates/mdk-core/src/encrypted_media/manager.rs]: I think this should be exported in the EncryptedMediaError enum, like `DecryptionFailedNoExporterSecretForEpoch(usize)`, so it can be matched programmatically

File: crates/mdk-core/src/encrypted_media/manager.rs [hunk 2]
```diff
@@ -1256,4 +1316,192 @@ mod tests {
...
+            .expect("Epoch hint decryption failed");
+        assert_eq!(
+            decrypted, data,
+            "Decrypted data should match original after epoch advancement"
+        );
+    }
+
+    #[test]
+    fn test_decrypt_from_download_epoch_hint() {
+        use crate::test_util::create_nostr_group_config_data;
```
Comment (erskingardner) [inline crates/mdk-core/src/encrypted_media/manager.rs]: Same here? let's move them to the top of the tests.

**Comment by erskingardner** [comment]
@JSKitty can you run `just precommit` to catch the formatting issue that's blocking CI?

**Comment by dannym-arx** [comment]
Awesome! Merged
